src/Cardano/BM/Tracer/Class.lhs

```diff
@@ -0,0 +1,59 @@
+\subsection{Cardano.BM.Tracer.Class}
+\label{code:Cardano.BM.Tracer.Class}
+
+\begin{code}
+module Cardano.BM.Tracer.Class
+    ( Tracer (..)
+    , Contravariant(..)
+    , nullTracer
+    , traceWith
+    ) where
+
+import           Data.Functor.Contravariant (Contravariant (..), Op (..))
+
+\end{code}
+
+The notion of a `Tracer` is an action that can be used to observe
+information of interest during evaluation. `Tracer`s can capture (and
+annotate) such observations with additional information from their
+execution context.
+
+\begin{code}
+newtype Tracer m s = Tracer { tracing :: Op (m ()) s }
+\end{code}
+
+A `Tracer` is an instance of `Contravariant`, which permits new
+`Tracer`s to be constructed that feed into the existing Tracer by use
+of `contramap`.
+
+\begin{code}
+instance Contravariant (Tracer m) where
+    contramap f = Tracer . contramap f . tracing
+
+\end{code}
+
+Although a `Tracer` is invoked in a monadic context (which may be
+`Identity`), the construction of a new Tracer is a pure function.
+
+This brings with it the constraint that the derived Tracer's form a
+hierachy which has it root at the top level tracer.
+
+The simplest tracer - one that does not generate output
+
+\begin{code}
+nullTracer :: (Applicative m) => Tracer m a
+nullTracer = Tracer $ Op $ \_ -> pure ()
+
+\end{code}
+
+
+Accepts a |Trace| and some payload |s|. First it gets the contravariant
+from the |Trace| as type "|Op (m ()) s|" and, after "|getOp :: b -> a|" which
+translates to "|s -> m ()|".
+\begin{code}
+traceWith :: Tracer m s -> s -> m ()
```

Comment:
maybe the names should be `tracingWith` - the idea here being that we are distinguishing this notion of "Trace" from other notions of "Trace" by appealing to it being in the present tense - ie you are _tracing with a tracer_